feat: add rule for missing 'provides' method in Deferrable ServiceProviders

PrinsFrank opened the pull request:
- [x] Added or updated tests
- [x] Documented user facing changes
- [x] Updated CHANGELOG.md

**Changes**

This rule will check for a missing 'provides' method in deferrable ServiceProviders.

The 'provides' method is present in the DeferrableProvider interface, but also in the abstract ServiceProvider. If it was not present in the abstract class, this check would not be necessary, but [that PR got rejected](https://github.com/laravel/framework/pull/42460) in Laravel.

**Breaking changes**

szepeviktor commented on src/Rules/DeferrableServiceProviderMissingProvidesRule.php:
```diff
@@ -0,0 +1,52 @@
+use PHPStan\Rules\RuleErrorBuilder;
+use PHPStan\ShouldNotHappenException;
+
+class DeferrableServiceProviderMissingProvidesRule implements Rule
```
Please add generics here: `@phpstan-implements ...`

szepeviktor commented on src/Rules/DeferrableServiceProviderMissingProvidesRule.php:
```diff
@@ -0,0 +1,52 @@
+    {
+        /** @var InClassNode $node */
+        $classReflection = $node->getClassReflection();
+        if ($classReflection->isSubclassOf(ServiceProvider::class) === false || $classReflection->implementsInterface(DeferrableProvider::class) === false) {
+            return []; // This rule is only applicable to deferrable serviceProviders
```
```suggestion
        // Only applicable to deferrable serviceProviders
        if ($classReflection->isSubclassOf(ServiceProvider::class) === false || $classReflection->implementsInterface(DeferrableProvider::class) === false) {
            return [];
```

szepeviktor commented on src/Rules/DeferrableServiceProviderMissingProvidesRule.php:
```diff
@@ -0,0 +1,52 @@
+        }
+
+        $method = $classReflection->getMethod('provides', new OutOfClassScope());
+        if ($method->getDeclaringClass()->getName() !== ServiceProvider::class) {
+            return []; // The provides method is overwritten somewhere in the class hierarchy
```
```suggestion
        // The `provides` method is overwritten somewhere in the class hierarchy
        if ($method->getDeclaringClass()->getName() !== ServiceProvider::class) {
            return [];
```

szepeviktor commented on src/Rules/DeferrableServiceProviderMissingProvidesRule.php:
```diff
@@ -0,0 +1,52 @@
+     */
+    public function processNode(Node $node, Scope $scope): array
+    {
+        /** @var InClassNode $node */
```
```suggestion
        /** @var \PHPStan\Node\InClassNode $node */
```

szepeviktor commented on src/Rules/DeferrableServiceProviderMissingProvidesRule.php:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    /**
+     * @throws ShouldNotHappenException
+     * @throws MissingMethodFromReflectionException
```
```suggestion
     * @throws \PHPStan\ShouldNotHappenException
     * @throws \PHPStan\Reflection\MissingMethodFromReflectionException
```

szepeviktor commented:
The problem is in https://github.com/nunomaduro/larastan/blob/master/tests/Rules/Data/relation-existence-rule.php

szepeviktor commented:
The rule name is also in `extension.neon`

PrinsFrank commented:
@szepeviktor I renamed the test case to confirm this is a test execution issue. That seems to indeed be the case..

PrinsFrank commented on tests/Rules/AbstractRuleTestCase.php:
```diff
@@ -0,0 +1,17 @@
+
+declare(strict_types=1);
+
+namespace Rules;
```
Tx! PHPStorm doesn't understand the namespaces in this project anymore, and neither do I really. Fixed!

szepeviktor commented on tests/Rules/CheckDispatchArgumentTypesCompatibleWithClassConstructorRuleTest.php:
```diff
@@ -9,10 +9,10 @@
 use PHPStan\Rules\FunctionCallParametersCheck;
 use PHPStan\Rules\Rule;
-use PHPStan\Testing\RuleTestCase;
+use Rules\AbstractRuleTestCase;
```
```suggestion
use Tests\Rules\AbstractRuleTestCase;
```

szepeviktor commented:
It seems to be green :)
🍏

szepeviktor commented:
What change in `phpunit.xml.dist` made tests pass?

szepeviktor commented on src/Rules/DeferrableServiceProviderMissingProvidesRule.php:
```diff
@@ -0,0 +1,56 @@
+        /** @var \PHPStan\Node\InClassNode $node */
+        $classReflection = $node->getClassReflection();
+        // This rule is only applicable to deferrable serviceProviders
+        if ($classReflection->isSubclassOf(ServiceProvider::class) === false || $classReflection->implementsInterface(DeferrableProvider::class) === false) {
+            return [];
+        }
+
+        if ($classReflection->hasNativeMethod('provides') === false) {
```
```suggestion
        if (!$classReflection->isSubclassOf(ServiceProvider::class) || !$classReflection->implementsInterface(DeferrableProvider::class)) {
            return [];
        }

        if (!$classReflection->hasNativeMethod('provides')) {
```

PrinsFrank commented:
> What change in `phpunit.xml.dist` made tests pass?

None, I added the beStrictAboutChangesToGlobalState setting to try to debug what was going on, but the issue ended up being in the way the container is booted and subsequently cached in the PHPStan RuleTestCase. All extensions need to be loaded or the next time the container is used it is missing dependencies. That's why I added the AbstractRuleTestCase.

PrinsFrank commented:
@szepeviktor @canvural @nunomaduro I'm cleaning up my forks, can I assume this is not going to get merged and close this PR?

PrinsFrank commented:
@szepeviktor Ok, np!

canvural commented:
Thank you! I made some changes to simplify it little bit.
